test/Microsoft.SqlTools.ServiceLayer.IntegrationTests/SqlProjects/SqlProjectsServiceTests.cs

```diff
@@ -157,6 +159,95 @@ await service.HandleDeleteSqlObjectScriptRequest(new SqlProjectScriptParams()
+            {
+                ProjectUri = projectUri,
+                SystemDatabase = SystemDatabase.MSDB,
+                DatabaseVariable = "$(EmEssDeeBee)",
+                SuppressMissingDependencies = false
+            }, requestMock.Object);
+
+            requestMock.AssertSuccess(nameof(service.HandleAddSystemDatabaseReferenceRequest));
+            Assert.AreEqual(1, service.Projects[projectUri].DatabaseReferences.Count, "Database references after adding system db reference");
+            SystemDatabaseReference systemDbRef = (SystemDatabaseReference)service.Projects[projectUri].DatabaseReferences.First(x => x is SystemDatabaseReference);
+            Assert.AreEqual(SystemDatabase.MSDB, systemDbRef.SystemDb, "Referenced system DB");
+            Assert.AreEqual("$(EmEssDeeBee)", systemDbRef.DatabaseVariable);
+            Assert.IsFalse(systemDbRef.SuppressMissingDependencies, nameof(systemDbRef.SuppressMissingDependencies));
+
+            // Validate adding a dacpac reference
+            string mockReferencePath = Path.Join(Path.GetDirectoryName(projectUri), "OtherDatabase.dacpac");
+
+            requestMock = new();
+            await service.HandleAddDacpacReferenceRequest(new AddDacpacReferenceParams()
+            {
+                ProjectUri = projectUri,
+                DacpacPath = mockReferencePath,
+                SuppressMissingDependencies = false,
+                DatabaseVariable = databaseVar.Name,
+                ServerVariable = serverVar.Name
+            }, requestMock.Object);
+
+            requestMock.AssertSuccess(nameof(service.HandleAddDacpacReferenceRequest));
+            Assert.AreEqual(2, service.Projects[projectUri].DatabaseReferences.Count, "Database references after adding dacpac reference");
+            DacpacReference dacpacRef = (DacpacReference)service.Projects[projectUri].DatabaseReferences.First(x => x is DacpacReference);
+            Assert.AreEqual(mockReferencePath, dacpacRef.DacpacPath, "Referenced dacpac");
+            Assert.AreEqual(databaseVar.Name, dacpacRef.DatabaseVariable);
+            Assert.AreEqual(serverVar.Name, dacpacRef.ServerVariable);
+            Assert.IsFalse(dacpacRef.SuppressMissingDependencies, nameof(dacpacRef.SuppressMissingDependencies));
+
+            // Validate adding a project reference
+            mockReferencePath = Path.Join(Path.GetDirectoryName(projectUri), "..", "OtherDatabase", "OtherDatabase.sqlproj");
+
+            requestMock = new();
+            await service.HandleAddSqlProjectReferenceRequest(new AddSqlProjectReferenceParams()
+            {
+                ProjectUri = projectUri,
+                ProjectPath = mockReferencePath,
+                ProjectGuid = TEST_GUID,
+                SuppressMissingDependencies = false,
+                DatabaseVariable = databaseVar.Name,
+                ServerVariable = serverVar.Name
```

Comment:
can one of the references in this test not have DatabaseVariable and ServerVariable to cover the "Same database" reference case?

Reply:
I was thinking about these tests as validating the unique parts of the Tools Service layer, so they'd need to make sure that all parameters are copied from the `Add____ReferenceParams` object to the constructor for the `____Reference` objects in their respective handlers.

Same database reference cases are handled and tested down in DacFx here, so I don't _think_ there's anything new to test up here.  Maybe I missed something, though?

https://msdata.visualstudio.com/SQLToolsAndLibraries/_git/DacFx?path=/Test/UnitTests/UTProjects/DatabaseReferenceTests.cs&version=GBmain&line=28&lineEnd=30&lineStartColumn=13&lineEndColumn=1&lineStyle=plain&_a=contents

Ofc, if you think it'd be better to cover those cases here as well, just let me know and I can add them!